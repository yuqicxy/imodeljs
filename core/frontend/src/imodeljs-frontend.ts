/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

export * from "./AccuDraw";
export * from "./AccuSnap";
export * from "./AuxCoordSys";
export * from "./CategorySelectorState";
export * from "./ContextRealityModelState";
export * from "./DevTools";
export * from "./DisplayStyleState";
export * from "./ElementLocateManager";
export * from "./EmphasizeElements";
export * from "./EntityState";
export * from "./FrontendLoggerCategory";
export * from "./FrontendRequestContext";
export * from "./FuzzySearch";
export * from "./GeoServices";
export * from "./HitDetail";
export * from "./IModelApp";
export * from "./IModelConnection";
export * from "./ImageUtil";
export * from "./Marker";
export * from "./ModelSelectorState";
export * from "./ModelState";
export * from "./NoRenderApp";
export * from "./NotificationManager";
export * from "./QuantityFormatter";
export * from "./RelativePosition";
export * from "./RenderCompatibility";
export * from "./RenderScheduleState";
export * from "./SelectionSet";
export * from "./Sheet";
export * from "./SolarCalculate";
export * from "./SpatialClassification";
export * from "./Sprites";
export * from "./StandardView";
export * from "./SubCategoriesCache";
export * from "./TentativePoint";
export * from "./TerrainProvider";
export * from "./ViewContext";
export * from "./ViewManager";
export * from "./ViewRect";
export * from "./ViewState";
export * from "./Viewport";
export * from "./oidc/OidcBrowserClient";
export * from "./oidc/OidcDesktopClientRenderer";
export * from "./plugin/Plugin";
export * from "./properties/AngleDescription";
export * from "./properties/BaseQuantityDescription";
export * from "./properties/Description";
export * from "./properties/EditorParams";
export * from "./properties/LengthDescription";
export * from "./properties/PrimitiveTypes";
export * from "./properties/Record";
export * from "./properties/ToolSettingsValue";
export * from "./properties/Value";
export * from "./render/FeatureSymbology";
export * from "./FeatureToggleClient";
export * from "./render/GraphicBuilder";
export * from "./render/MockRender";
export * from "./render/System";
export * from "./render/webgl/PerformanceMetrics";
export * from "./render/webgl/Target";
export * from "./tile/A3xTile";
export * from "./tile/A3xTileIO";
export * from "./tile/BackgroundTerrainTileTree";
export * from "./tile/BingElevation";
export * from "./tile/CesiumWorldTerrainTileTree";
export * from "./tile/GltfReader";
export * from "./tile/IModelTileLoader";
export * from "./tile/ImdlReader";
export * from "./tile/Tile";
export * from "./tile/TileAdmin";
export * from "./tile/TileRequest";
export * from "./tile/TileTree";
export * from "./tile/WebMapTileTree";
export * from "./tools/AccuDrawTool";
export * from "./tools/ClipViewTool";
export * from "./tools/EditManipulator";
export * from "./tools/ElementSetTool";
export * from "./tools/EventController";
export * from "./tools/IdleTool";
export * from "./tools/MeasureTool";
export * from "./tools/PrimitiveTool";
export * from "./tools/SelectTool";
export * from "./tools/Tool";
export * from "./tools/ToolSettings";
export * from "./tools/ToolAdmin";
export * from "./tools/ToolAssistance";
export * from "./tools/ViewTool";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
declare var BUILD_TYPE: string;
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("imodeljs-frontend", BUILD_SEMVER);
  (window as any).iModelJsVersions.set("buildType", BUILD_TYPE);
}

/** @docs-package-description
 * The imodeljs-frontend package always runs in a web browser. It contains classes for [querying iModels and showing views]($docs/learning/frontend/index.md).
 */

/**
 * @docs-group-description IModelApp
 * Classes for configuring and administering an iModel.js application.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description IModelConnection
 * Classes for working with a connection to an [iModel briefcase]($docs/learning/IModels.md)
 */
/**
 * @docs-group-description ElementState
 * Classes for working with the *state* of Elements in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description ModelState
 * Classes for working with the *state* of Models in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Tools
 * Classes for [working with Tools]($docs/learning/frontend/Tools.md)
 */
/**
 * @docs-group-description Views
 * Classes for [working with Views]($docs/learning/frontend/Views.md)
 */
/**
 * @docs-group-description LocatingElements
 * Classes for locating and snapping to elements in views.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description AccuDraw
 * AccuDraw provides helpful assistance for creating and modifying elements in a view.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Notifications
 * Notifications provide feedback to the user of something of interest.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Plugins
 * Classes for creating and managing runtime [Plugins]($docs/learning/frontend/Plugins.md)
 */
/**
 * @docs-group-description Properties
 * Classes for working with property records and descriptions.
 */
/**
 * @docs-group-description Rendering
 * Classes for rendering the contents of views.
 */
/**
 * @docs-group-description SelectionSet
 * Classes for working with the set of selected elements.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description OIDC
 * Classes for working with the OpenID Connect (OIDC) protocol
 */
/**
 * @docs-group-description Utils
 * Miscellaneous utility classes.
 */
/**
 * @docs-group-description Logging
 * Logger categories used by this package
 */
/**
 * @docs-group-description SpatialClassification
 * Classes for spatial classification.
 */
/**
 * @docs-group-description Features
 * Classes for Feature Tracking classification.
 */
