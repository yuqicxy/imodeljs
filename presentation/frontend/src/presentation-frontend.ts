/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/**
 * @module Core
 *
 * @docs-group-description Core
 * Common types used for retrieving presentation data from iModels.
 */
export { PersistenceHelper } from "./PersistenceHelper";
export { Presentation } from "./Presentation";
export { PresentationManager, PresentationManagerProps } from "./PresentationManager";
export { RulesetManager } from "./RulesetManager";
export { RulesetVariablesManager } from "./RulesetVariablesManager";
export { FavoritePropertiesManager, FavoriteProperties } from "./favorite-properties/FavoritePropertiesManager";
export { IFavoritePropertiesStorage, IModelAppFavoritePropertiesStorage } from "./favorite-properties/FavoritePropertiesStorage";

/**
 * @module UnifiedSelection
 *
 * @docs-group-description UnifiedSelection
 * Types related to [unified selection]($docs/learning/presentation/Unified-Selection/index.md).
 */
export { SelectionChangeEvent, SelectionChangeEventArgs, SelectionChangeType, SelectionChangesListener } from "./selection/SelectionChangeEvent";
export { ISelectionProvider } from "./selection/ISelectionProvider";
export { SelectionManager, SelectionManagerProps } from "./selection/SelectionManager";
export { SelectionScopesManager, SelectionScopesManagerProps, getScopeId } from "./selection/SelectionScopesManager";
export { SelectionHandler } from "./selection/SelectionHandler";
export { HiliteSet, HiliteSetProvider } from "./selection/HiliteSetProvider";
export { SelectionHelper } from "./selection/SelectionHelper";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
/* istanbul ignore next */
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("presentation-frontend", BUILD_SEMVER);
}
