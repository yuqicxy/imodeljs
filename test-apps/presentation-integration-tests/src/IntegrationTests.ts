/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
// tslint:disable:no-direct-imports
import * as fs from "fs";
import * as path from "path";
import * as cpx from "cpx";
import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
import { I18NOptions } from "@bentley/imodeljs-i18n";
import { Logger, LogLevel } from "@bentley/bentleyjs-core";
import { LoggingNamespaces, RequestPriority } from "@bentley/presentation-common";
import { PresentationProps as PresentationBackendProps } from "@bentley/presentation-backend";
import { PresentationManagerProps as PresentationFrontendProps } from "@bentley/presentation-frontend";
import { NoRenderApp, IModelAppOptions } from "@bentley/imodeljs-frontend";
import { initializeAsync as initializeTesting, terminate as terminateTesting, PresentationTestingInitProps } from "@bentley/presentation-testing";
import { IModelJsConfig } from "@bentley/config-loader/lib/IModelJsConfig";

process.env.NODE_ENV = "development";

IModelJsConfig.init(true);

const copyBentleyBackendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "assets");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

const copyBentleyFrontendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "public");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

class IntegrationTestsApp extends NoRenderApp {
  protected static supplyI18NOptions(): I18NOptions {
    const urlTemplate = "file://" + path.join(path.resolve("lib/public/locales"), "{{lng}}/{{ns}}.json").replace(/\\/g, "/");
    return { urlTemplate };
  }

  public static startup(opts?: IModelAppOptions) {
    NoRenderApp.startup({ ...opts, i18n: this.supplyI18NOptions() });
    cpx.copySync(`assets/**/*`, "lib/assets");
    copyBentleyBackendAssets("lib/assets");
    copyBentleyFrontendAssets("lib/public");
  }
}

export const initialize = async (backendTimeout: number = 0) => {
  // init logging
  Logger.initializeToConsole();
  Logger.setLevelDefault(LogLevel.Error);
  Logger.setLevel(LoggingNamespaces.ECObjects_ECExpressions, LogLevel.Warning);
  Logger.setLevel(LoggingNamespaces.ECPresentation, LogLevel.Warning);

  const backendInitProps: PresentationBackendProps = {
    requestTimeout: backendTimeout,
    rulesetDirectories: ["lib/assets/rulesets"],
    localeDirectories: ["lib/assets/locales"],
    activeLocale: "en-PSEUDO",
    taskAllocationsMap: {
      [RequestPriority.Max]: 1,
    },
  };
  const frontendInitProps: PresentationFrontendProps = {
    activeLocale: "en-PSEUDO",
  };

  const presentationTestingInitProps: PresentationTestingInitProps = {
    backendProps: backendInitProps,
    frontendProps: frontendInitProps,
    frontendApp: IntegrationTestsApp,
    useClientServices: true,
  };

  await initializeTesting(presentationTestingInitProps);
};

export const terminate = () => {
  terminateTesting();
};
