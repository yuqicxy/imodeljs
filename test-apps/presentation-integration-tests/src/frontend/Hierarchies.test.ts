/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as faker from "faker";
import { initialize, terminate } from "../IntegrationTests";
import { Id64, using } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import {
  InstanceKey, Ruleset, RuleTypes, ChildNodeSpecificationTypes,
  KeySet, ECInstanceNodeKey, getInstancesCount, RegisteredRuleset,
} from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";

describe("Hierarchies", () => {

  let imodel: IModelConnection;

  before(async () => {
    await initialize();
    const testIModelName: string = "assets/datasets/Properties_60InstancesWithUrl2.ibim";
    imodel = await IModelConnection.openSnapshot(testIModelName);
    expect(imodel).is.not.null;
  });

  after(async () => {
    await imodel.closeSnapshot();
    terminate();
  });

  describe("NodesPaths", () => {

    it("gets filtered node paths", async () => {
      const ruleset: Ruleset = require("../../test-rulesets/NodePaths/getFilteredNodePaths");
      /* Hierarchy in the ruleset:
      filter r1
        filter ch1
        other ch2
        other ch3
          filter ch4
      other r2
      other r3
        other ch5
        filter ch6
      */
      await using(await Presentation.presentation.rulesets().add(ruleset), async (_r) => {
        const result = await Presentation.presentation.getFilteredNodePaths({ imodel, rulesetId: ruleset.id }, "filter");
        expect(result).to.matchSnapshot();
      });
    });

    it("gets node paths", async () => {
      const ruleset: Ruleset = require("../../test-rulesets/NodePaths/getNodePaths");
      /*
      [BisCore:RepositoryModel] 0x1
        [BisCore:Subject] 0x1
          [BisCore:DefinitionPartition] ECClassGroupingNode
            [BisCore:DefinitionPartition] 0x10
          [BisCore:LinkPartition] ECClassGroupingNode
            [BisCore:LinkPartition] 0xe
      */
      await using<RegisteredRuleset, Promise<void>>(await Presentation.presentation.rulesets().add(ruleset), async () => {
        const key1: InstanceKey = { id: Id64.fromString("0x1"), className: "BisCore:RepositoryModel" };
        const key2: InstanceKey = { id: Id64.fromString("0x1"), className: "BisCore:Subject" };
        const key3: InstanceKey = { id: Id64.fromString("0x10"), className: "BisCore:DefinitionPartition" };
        const key4: InstanceKey = { id: Id64.fromString("0xe"), className: "BisCore:LinkPartition" };
        const keys: InstanceKey[][] = [[key1, key2, key3], [key1, key2, key4]];

        const result = await Presentation.presentation.getNodePaths({ imodel, rulesetId: ruleset.id }, keys, 1);
        expect(result).to.matchSnapshot();
      });

    });

  });

  describe("Counting instances of selected nodes", () => {

    it("correctly counts instances when key set contains grouping node keys", async () => {
      const ruleset: Ruleset = {
        id: faker.random.word(),
        rules: [{
          ruleType: RuleTypes.RootNodes,
          specifications: [{
            specType: ChildNodeSpecificationTypes.InstanceNodesOfSpecificClasses,
            classes: { schemaName: "BisCore", classNames: ["Model"] },
            arePolymorphic: true,
            groupByClass: true,
            groupByLabel: false,
          }],
        }],
      };
      await using<RegisteredRuleset, Promise<void>>(await Presentation.presentation.rulesets().add(ruleset), async () => {
        const rootNodes = await Presentation.presentation.getNodes({ imodel, rulesetId: ruleset.id });
        expect(rootNodes).to.matchSnapshot();
        /*
        The result should look like this (all grouping nodes):
          Label                 Grouped Instances Count
          Definition Model      1
          Dictionary Model      1
          Document List         2
          Group Model           1
          Link Model            1
          Physical Model        1
          Repository Model      1

        we're going to count instances for:
          - one of the definition model node keys
          - dictionary model's instance key
          - document list grouping node key
        the result should be 1 + 1 + 2 = 4
        */

        const definitionModelNodes = await Presentation.presentation.getNodes(
          { imodel, rulesetId: ruleset.id }, rootNodes[0].key);
        const dictionaryModelNodes = await Presentation.presentation.getNodes(
          { imodel, rulesetId: ruleset.id }, rootNodes[1].key);

        const keys = new KeySet([
          definitionModelNodes[0].key,
          (dictionaryModelNodes[0].key as ECInstanceNodeKey).instanceKey,
          rootNodes[2].key,
        ]);
        const instancesCount = getInstancesCount(keys);
        expect(instancesCount).to.eq(4);
      });
    });

  });

});
