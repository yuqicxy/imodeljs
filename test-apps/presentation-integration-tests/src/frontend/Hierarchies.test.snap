// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hierarchies Counting instances of selected nodes correctly counts instances when key set contains grouping node keys 1`] = `
Array [
  Object {
    "description": "A container for persisting definition elements.",
    "hasChildren": true,
    "key": Object {
      "className": "BisCore:DefinitionModel",
      "groupedInstancesCount": 1,
      "pathFromRoot": Array [
        "851bf7b261cc7fe7421475441e4e1b20",
      ],
      "type": "ECClassGroupingNode",
    },
    "label": "Definition Model",
    "labelDefinition": Object {
      "displayValue": "Definition Model",
      "rawValue": "Definition Model",
      "typeName": "string",
    },
  },
  Object {
    "description": "The singleton container for repository-specific definition elements.",
    "hasChildren": true,
    "key": Object {
      "className": "BisCore:DictionaryModel",
      "groupedInstancesCount": 1,
      "pathFromRoot": Array [
        "96989c2a849b374d693e051e2e99a1f2",
      ],
      "type": "ECClassGroupingNode",
    },
    "label": "Dictionary Model",
    "labelDefinition": Object {
      "displayValue": "Dictionary Model",
      "rawValue": "Dictionary Model",
      "typeName": "string",
    },
  },
  Object {
    "description": "Contains a list of document elements.",
    "hasChildren": true,
    "key": Object {
      "className": "BisCore:DocumentListModel",
      "groupedInstancesCount": 2,
      "pathFromRoot": Array [
        "8919efb29e1b6e218bb4254cdf8a53c6",
      ],
      "type": "ECClassGroupingNode",
    },
    "label": "Document List",
    "labelDefinition": Object {
      "displayValue": "Document List",
      "rawValue": "Document List",
      "typeName": "string",
    },
  },
  Object {
    "hasChildren": true,
    "key": Object {
      "className": "Generic:GroupModel",
      "groupedInstancesCount": 1,
      "pathFromRoot": Array [
        "6262de971bde6729bd04cecec9d7b802",
      ],
      "type": "ECClassGroupingNode",
    },
    "label": "Group Model",
    "labelDefinition": Object {
      "displayValue": "Group Model",
      "rawValue": "Group Model",
      "typeName": "string",
    },
  },
  Object {
    "description": "A container for persisting link elements.",
    "hasChildren": true,
    "key": Object {
      "className": "BisCore:LinkModel",
      "groupedInstancesCount": 1,
      "pathFromRoot": Array [
        "70e01c48fea080390154e5920a8025e8",
      ],
      "type": "ECClassGroupingNode",
    },
    "label": "Link Model",
    "labelDefinition": Object {
      "displayValue": "Link Model",
      "rawValue": "Link Model",
      "typeName": "string",
    },
  },
  Object {
    "description": "A container for persisting physical elements that model physical space.",
    "hasChildren": true,
    "key": Object {
      "className": "BisCore:PhysicalModel",
      "groupedInstancesCount": 1,
      "pathFromRoot": Array [
        "92a67755a54843895852bfb646db1fd1",
      ],
      "type": "ECClassGroupingNode",
    },
    "label": "Physical Model",
    "labelDefinition": Object {
      "displayValue": "Physical Model",
      "rawValue": "Physical Model",
      "typeName": "string",
    },
  },
  Object {
    "description": "The singleton container of repository-related information elements.",
    "hasChildren": true,
    "key": Object {
      "className": "BisCore:RepositoryModel",
      "groupedInstancesCount": 1,
      "pathFromRoot": Array [
        "61beb3c378f7696c0ede9a111ed149d3",
      ],
      "type": "ECClassGroupingNode",
    },
    "label": "Repository Model",
    "labelDefinition": Object {
      "displayValue": "Repository Model",
      "rawValue": "Repository Model",
      "typeName": "string",
    },
  },
]
`;

exports[`Hierarchies NodesPaths gets filtered node paths 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "filteringData": Object {
          "childMatchesCount": 0,
          "matchesCount": 1,
        },
        "index": 0,
        "isMarked": false,
        "node": Object {
          "description": "description",
          "imageId": "imageId",
          "key": Object {
            "pathFromRoot": Array [
              "247eca1de00df6e49f5467ee69db2f91",
              "86bc2e93bbfcfa542836ba8ede7b3033",
            ],
            "type": "nodeType",
          },
          "label": "filter ch1",
          "labelDefinition": Object {
            "displayValue": "filter ch1",
            "rawValue": "filter ch1",
            "typeName": "string",
          },
        },
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "filteringData": Object {
              "childMatchesCount": 0,
              "matchesCount": 1,
            },
            "index": 0,
            "isMarked": false,
            "node": Object {
              "description": "description",
              "imageId": "imageId",
              "key": Object {
                "pathFromRoot": Array [
                  "247eca1de00df6e49f5467ee69db2f91",
                  "a1fcb2fc6fd92f330d9f11685a4cdab5",
                  "0f4517b8731a62e65c5d1ce1680a3df1",
                ],
                "type": "nodeType",
              },
              "label": "filter ch4",
              "labelDefinition": Object {
                "displayValue": "filter ch4",
                "rawValue": "filter ch4",
                "typeName": "string",
              },
            },
          },
        ],
        "filteringData": Object {
          "childMatchesCount": 1,
          "matchesCount": 0,
        },
        "index": 1,
        "isMarked": false,
        "node": Object {
          "description": "description",
          "hasChildren": true,
          "imageId": "imageId",
          "key": Object {
            "pathFromRoot": Array [
              "247eca1de00df6e49f5467ee69db2f91",
              "a1fcb2fc6fd92f330d9f11685a4cdab5",
            ],
            "type": "nodeType",
          },
          "label": "other ch3",
          "labelDefinition": Object {
            "displayValue": "other ch3",
            "rawValue": "other ch3",
            "typeName": "string",
          },
        },
      },
    ],
    "filteringData": Object {
      "childMatchesCount": 2,
      "matchesCount": 1,
    },
    "index": 0,
    "isMarked": false,
    "node": Object {
      "description": "description",
      "hasChildren": true,
      "imageId": "imageId",
      "key": Object {
        "pathFromRoot": Array [
          "247eca1de00df6e49f5467ee69db2f91",
        ],
        "type": "nodeType",
      },
      "label": "filter r1",
      "labelDefinition": Object {
        "displayValue": "filter r1",
        "rawValue": "filter r1",
        "typeName": "string",
      },
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "filteringData": Object {
          "childMatchesCount": 0,
          "matchesCount": 1,
        },
        "index": 0,
        "isMarked": false,
        "node": Object {
          "description": "description",
          "imageId": "imageId",
          "key": Object {
            "pathFromRoot": Array [
              "fc89c0af4376bb60df14f648618e822d",
              "06af315ac469b705ed23f5675daaff5a",
            ],
            "type": "nodeType",
          },
          "label": "filter ch6",
          "labelDefinition": Object {
            "displayValue": "filter ch6",
            "rawValue": "filter ch6",
            "typeName": "string",
          },
        },
      },
    ],
    "filteringData": Object {
      "childMatchesCount": 1,
      "matchesCount": 0,
    },
    "index": 1,
    "isMarked": false,
    "node": Object {
      "description": "description",
      "hasChildren": true,
      "imageId": "imageId",
      "key": Object {
        "pathFromRoot": Array [
          "fc89c0af4376bb60df14f648618e822d",
        ],
        "type": "nodeType",
      },
      "label": "other r3",
      "labelDefinition": Object {
        "displayValue": "other r3",
        "rawValue": "other r3",
        "typeName": "string",
      },
    },
  },
]
`;

exports[`Hierarchies NodesPaths gets node paths 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "filteringData": Object {
                  "childMatchesCount": 0,
                  "matchesCount": 0,
                },
                "index": 0,
                "isMarked": false,
                "node": Object {
                  "key": Object {
                    "instanceKey": Object {
                      "className": "BisCore:DefinitionPartition",
                      "id": "0x10",
                    },
                    "instanceKeys": Array [
                      Object {
                        "className": "BisCore:DefinitionPartition",
                        "id": "0x10",
                      },
                    ],
                    "pathFromRoot": Array [
                      "679ccbb46ad76244084d352b5d7f7983",
                      "2711094066a9995f601d286b6e6866e6",
                      "44d4d5acfd7d3656dcb40a5281d67c61",
                      "8b9e160f3866c7f4fadae569eba9ef25",
                      "0efd370a8754de2427d0fd20b7714777",
                    ],
                    "type": "ECInstancesNode",
                  },
                  "label": "BisCore.DictionaryModel",
                  "labelDefinition": Object {
                    "displayValue": "BisCore.DictionaryModel",
                    "rawValue": "BisCore.DictionaryModel",
                    "typeName": "string",
                  },
                },
              },
            ],
            "filteringData": Object {
              "childMatchesCount": 0,
              "matchesCount": 0,
            },
            "index": 0,
            "isMarked": false,
            "node": Object {
              "description": "A Definition Partition element indicates that there is a definition-related modeling perspective within the overall DgnDb information hierarchy. A Definition Partition is always parented to a Subject and broken down by a Definition Model.",
              "hasChildren": true,
              "key": Object {
                "className": "BisCore:DefinitionPartition",
                "groupedInstancesCount": 1,
                "pathFromRoot": Array [
                  "679ccbb46ad76244084d352b5d7f7983",
                  "2711094066a9995f601d286b6e6866e6",
                  "44d4d5acfd7d3656dcb40a5281d67c61",
                  "8b9e160f3866c7f4fadae569eba9ef25",
                ],
                "type": "ECClassGroupingNode",
              },
              "label": "Definition Partition",
              "labelDefinition": Object {
                "displayValue": "Definition Partition",
                "rawValue": "Definition Partition",
                "typeName": "string",
              },
            },
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "filteringData": Object {
                  "childMatchesCount": 0,
                  "matchesCount": 0,
                },
                "index": 0,
                "isMarked": true,
                "node": Object {
                  "key": Object {
                    "instanceKey": Object {
                      "className": "BisCore:LinkPartition",
                      "id": "0xe",
                    },
                    "instanceKeys": Array [
                      Object {
                        "className": "BisCore:LinkPartition",
                        "id": "0xe",
                      },
                    ],
                    "pathFromRoot": Array [
                      "679ccbb46ad76244084d352b5d7f7983",
                      "2711094066a9995f601d286b6e6866e6",
                      "44d4d5acfd7d3656dcb40a5281d67c61",
                      "93e5eef1bcd27c39e00b3f62d4df415b",
                      "1554b13b284ed9a3ac984f4d0b885cfc",
                    ],
                    "type": "ECInstancesNode",
                  },
                  "label": "BisCore.RealityDataSources",
                  "labelDefinition": Object {
                    "displayValue": "BisCore.RealityDataSources",
                    "rawValue": "BisCore.RealityDataSources",
                    "typeName": "string",
                  },
                },
              },
            ],
            "filteringData": Object {
              "childMatchesCount": 0,
              "matchesCount": 0,
            },
            "index": 1,
            "isMarked": false,
            "node": Object {
              "description": "A Link Partition element indicates that there is a link-related modeling perspective within the overall DgnDb information hierarchy. A Link Partition is always parented to a Subject and broken down by a Link Model.",
              "hasChildren": true,
              "key": Object {
                "className": "BisCore:LinkPartition",
                "groupedInstancesCount": 1,
                "pathFromRoot": Array [
                  "679ccbb46ad76244084d352b5d7f7983",
                  "2711094066a9995f601d286b6e6866e6",
                  "44d4d5acfd7d3656dcb40a5281d67c61",
                  "93e5eef1bcd27c39e00b3f62d4df415b",
                ],
                "type": "ECClassGroupingNode",
              },
              "label": "Link Partition",
              "labelDefinition": Object {
                "displayValue": "Link Partition",
                "rawValue": "Link Partition",
                "typeName": "string",
              },
            },
          },
        ],
        "filteringData": Object {
          "childMatchesCount": 0,
          "matchesCount": 0,
        },
        "index": 0,
        "isMarked": false,
        "node": Object {
          "hasChildren": true,
          "key": Object {
            "instanceKey": Object {
              "className": "BisCore:Subject",
              "id": "0x1",
            },
            "instanceKeys": Array [
              Object {
                "className": "BisCore:Subject",
                "id": "0x1",
              },
            ],
            "pathFromRoot": Array [
              "679ccbb46ad76244084d352b5d7f7983",
              "2711094066a9995f601d286b6e6866e6",
              "44d4d5acfd7d3656dcb40a5281d67c61",
            ],
            "type": "ECInstancesNode",
          },
          "label": "DgnV8Bridge",
          "labelDefinition": Object {
            "displayValue": "DgnV8Bridge",
            "rawValue": "DgnV8Bridge",
            "typeName": "string",
          },
        },
      },
    ],
    "filteringData": Object {
      "childMatchesCount": 0,
      "matchesCount": 0,
    },
    "index": 0,
    "isMarked": false,
    "node": Object {
      "hasChildren": true,
      "key": Object {
        "instanceKey": Object {
          "className": "BisCore:RepositoryModel",
          "id": "0x1",
        },
        "instanceKeys": Array [
          Object {
            "className": "BisCore:RepositoryModel",
            "id": "0x1",
          },
        ],
        "pathFromRoot": Array [
          "679ccbb46ad76244084d352b5d7f7983",
          "2711094066a9995f601d286b6e6866e6",
        ],
        "type": "ECInstancesNode",
      },
      "label": "DgnV8Bridge",
      "labelDefinition": Object {
        "displayValue": "DgnV8Bridge",
        "rawValue": "DgnV8Bridge",
        "typeName": "string",
      },
    },
  },
]
`;
